hikari/interactions/modal_interactions.py

```diff
@@ -0,0 +1,359 @@
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""Models and enums used for Discord's Modals interaction flow."""
+
+from __future__ import annotations
+
+__all__: typing.List[str] = [
+    "ModalResponseTypesT",
+    "ModalInteraction",
+    "InteractionTextInput",
+    "ModalInteraction",
+    "TextInputStyle",
+    "ModalComponentType",
+    "PartialModalComponent",
+    "ModalActionRowComponent",
+]
+
+import typing
+
+import attr
+
+from hikari import channels
+from hikari import guilds
+from hikari import messages
+from hikari import permissions
+from hikari import snowflakes
+from hikari import traits
+from hikari.interactions import base_interactions
+from hikari.internal import attr_extensions
+from hikari.internal import enums
+
+if typing.TYPE_CHECKING:
+    from hikari import users as _users
+    from hikari.api import special_endpoints
+
+ModalResponseTypesT = typing.Literal[
+    base_interactions.ResponseType.MESSAGE_CREATE, 4, base_interactions.ResponseType.DEFERRED_MESSAGE_CREATE, 5
+]
+"""Type-hint of the response types which are valid for a modal interaction.
+
+The following types are valid for this:
+
+* `hikari.interactions.base_interactions.ResponseType.MESSAGE_CREATE`/`4`
+* `hikari.interactions.base_interactions.ResponseType.DEFERRED_MESSAGE_CREATE`/`5`
+"""
+
+
+@typing.final
+class ModalComponentType(int, enums.Enum):
+    """Types of components found within Discord."""
+
+    ACTION_ROW = 1
+    """A non-interactive container component for other types of components.
+
+    !!! note
+        As this is a container component it can never be contained within another
+        component and therefore will always be top-level.
+
+    !!! note
+        As of writing this can only contain one component type.
+    """
+
+    TEXT_INPUT = 4
+    """A text input component
+
+    !! note
+        This component may only be used inside a modal container.
+    """
+
+
+class TextInputStyle(int, enums.Enum):
+    """A text input style."""
+
+    SHORT = 1
+    """Intended for short single-line text."""
+
+    PARAGRAPH = 2
+    """Intended for much longer inputs."""
+
+
+@attr.define(kw_only=True, weakref_slot=False)
+class PartialModalComponent:
+    """Base class for all model component entities."""
+
+    type: typing.Union[ModalComponentType, int] = attr.field()
+    """The type of component this is."""
+
+
+@attr.define(weakref_slot=False)
+class ModalActionRowComponent(PartialModalComponent):
+    """Represents a row of components attached to a message.
+
+    !!! note
+        This is a top-level container component and will never be found within
+        another component.
+    """
+
+    components: typing.Sequence[PartialModalComponent] = attr.field()
```

Comment:
In practice, this will be filled by `InteractionTextInput` right?

Why not having it typed as typing.Sequence[InteractionTextInput] then?

Reply:
Same as above

Reply:
Make sense in this case